server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalVariableWidthHistogram.java

```diff
@@ -0,0 +1,511 @@
+        Bucket merged_bucket = to_merge.get(0).reduce(to_merge, reduceContext);
+        buckets.set(idx1, merged_bucket);
+        buckets.remove(idx2);
+    }
+
+    /**
+     * This method should not be called for this specific subclass of InternalHistogram, since there should not be
+     * empty buckets when clustering.
+=    */
+    private double nextKey(double key){ return key + 1; }
+
+    private static class IteratorAndCurrent {
+
+        private final Iterator<Bucket> iterator;
+        private Bucket current;
+
+        IteratorAndCurrent(Iterator<Bucket> iterator) {
+            this.iterator = iterator;
+            current = iterator.next();
+        }
+
+    }
+
+    private List<Bucket> reduceBuckets(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
+        // Use a min heap (priority queue) to merge the lists of buckets together
+        final PriorityQueue<IteratorAndCurrent> pq = new PriorityQueue<IteratorAndCurrent>(aggregations.size()) {
+            @Override
+            protected boolean lessThan(IteratorAndCurrent a, IteratorAndCurrent b) {
+                return Double.compare(a.current.centroid, b.current.centroid) < 0;
+            }
+        };
+        for (InternalAggregation aggregation : aggregations) {
+            InternalVariableWidthHistogram histogram = (InternalVariableWidthHistogram) aggregation;
+            if (histogram.buckets.isEmpty() == false) {
+                pq.add(new IteratorAndCurrent(histogram.buckets.iterator()));
+            }
+        }
+
+        List<Bucket> reducedBuckets = new ArrayList<>();
+        while (pq.size() > 0){
+            final IteratorAndCurrent top = pq.top();
+            reducedBuckets.add(top.current);
+
+            if(reduceContext.isFinalReduce() == false){
+                reduceContext.consumeBucketsAndMaybeBreak(1);
+            } else{
+                reduceContext.consumeBucketsAndMaybeBreak(-countInnerBucket(top.current));
+            }
+
+            if (top.iterator.hasNext()) {
+                final Bucket next = top.iterator.next();
+                assert next.compareKey(top.current) >= 0: "shards must return data sorted by key";
+                top.current = next;
+                pq.updateTop();
+            } else {
+                pq.pop();
+            }
+        }
+
+        mergeBucketsIfNeeded(reducedBuckets, reduceContext);
+        adjustMinBoundsIfOverlap(reducedBuckets);
+        mergeBucketsWidthIdenticalMin(reducedBuckets, reduceContext);
+
+        return reducedBuckets;
+    }
+
+    /**
+     * Merges closest buckets together until <code>reducedBuckets</code> has <code>numClusters</code> buckets.
+     * Requires: <code>reducedBuckets</code> is sorted in ascending order by bucket centroid.
+     */
+    private void mergeBucketsIfNeeded(List<Bucket> reducedBuckets, ReduceContext reduceContext){
+        while(reducedBuckets.size() > numClusters){
```

Comment:
Ditto to this loop.  Although tbf I only skimmed the actual merging, so it might be provable that they exit in a few iterations, in which case this might not be a concern. :)